### ISSUE TITLE
Agent doesn't exit when SSH session ends

### ISSUE THREAD

**Issue by opie4624** [romanz/trezor-agent#6]
```
 ⚙ username@workstation > ~ > trezor-agent -vvvvv -c username@server.example.com
2016-01-05 15:11:55,546 DEBUG        connected to Trezor 358EB5A28279FB09FDA78CE9                                                         [client.py:23]
2016-01-05 15:11:55,546 DEBUG        label    : Trezor                                                                                    [client.py:24]
2016-01-05 15:11:55,546 DEBUG        vendor   : bitcointrezor.com                                                                         [client.py:25]
2016-01-05 15:11:55,546 DEBUG        version  : 1.3.4                                                                                     [client.py:28]
2016-01-05 15:11:55,546 DEBUG        revision : db93a50f76204418a2cf7d2c7e0391f486729bf3                                                  [client.py:29]
2016-01-05 15:11:55,560 DEBUG        parsed identity: {'path': None, 'host': 'server.example.com', 'proto': None, 'port': None, 'user': 'username'}  [client.py:101]
2016-01-05 15:11:55,561 INFO         getting "ssh://username@server.example.com" public key (nist256p1) from Trezor...                               [client.py:54]
2016-01-05 15:11:55,561 DEBUG        address string: '\x00\x00\x00\x00ssh://username@server.example.com'                                             [client.py:123]
Use the numeric keypad to describe number positions. The layout is:
    7 8 9
    4 5 6
    1 2 3
Please enter current PIN:

2016-01-05 15:12:19,422 DEBUG        fingerprint: 25:ef:06:74:6a:3a:fb:d2:3e:d1:99:aa:c9:07:da:78                                         [formats.py:141]
2016-01-05 15:12:19,422 DEBUG        parsed identity: {'path': None, 'host': 'server.example.com', 'proto': None, 'port': None, 'user': 'username'}  [client.py:101]
2016-01-05 15:12:19,422 DEBUG        SSH connect: ['ssh', '-l', 'username', 'server.example.com']                                                    [__main__.py:109]
2016-01-05 15:12:19,425 DEBUG        loading SSH public key: 'ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBBgVDxlnq7abeYYL4Ga96ONAhFkznTm0z824iV7D2euJKZpf5U+pTtsjhxgiVKE8hHjTD9bD/PNp8wXNGwt3T7E= ssh://username@server.example.com\n' [formats.py:148]
2016-01-05 15:12:19,426 DEBUG        key type: ecdsa-sha2-nistp256                                                                        [formats.py:37]
2016-01-05 15:12:19,426 DEBUG        curve name: nistp256                                                                                 [formats.py:44]
2016-01-05 15:12:19,426 DEBUG        loaded ecdsa-sha2-nistp256 public key: 25:ef:06:74:6a:3a:fb:d2:3e:d1:99:aa:c9:07:da:78               [formats.py:154]
2016-01-05 15:12:19,426 DEBUG        serving on SSH_AUTH_SOCK=/var/folders/s5/1f6r4kbj6v70f3dp96n8bm9r00056p/T/ssh-agent-UH1bVZ           [server.py:26]
2016-01-05 15:12:19,426 DEBUG        server thread started                                                                                [server.py:53]
2016-01-05 15:12:19,426 DEBUG        running ['ssh', '-l', 'username', 'server.example.com'] with {'SSH_AUTH_SOCK': '/var/folders/s5/1f6r4kbj6v70f3dp96n8bm9r00056p/T/ssh-agent-UH1bVZ', 'SSH_AGENT_PID': '5533'} [server.py:92]
2016-01-05 15:12:19,426 DEBUG        waiting for connection on /var/folders/s5/1f6r4kbj6v70f3dp96n8bm9r00056p/T/ssh-agent-UH1bVZ          [server.py:55]
2016-01-05 15:12:19,429 DEBUG        subprocess 5543 is running                                                                           [server.py:99]
[detached (from session 0)]
 username@server > ~ >
Shared connection to server.example.com closed.
2016-01-05 15:13:50,986 DEBUG        subprocess 5543 exited: 0                                                                            [server.py:101]
2016-01-05 15:13:50,986 DEBUG        closing server                                                                                       [server.py:87]
2016-01-05 15:13:50,987 INFO         disconnected from Trezor                                                                             [client.py:41]
Traceback (most recent call last):
  File "/usr/local/bin/trezor-agent", line 11, in <module>
    sys.exit(trezor_agent())
  File "/usr/local/lib/python2.7/site-packages/trezor_agent/__main__.py", line 129, in trezor_agent
    run_agent(trezor.Client)
  File "/usr/local/lib/python2.7/site-packages/trezor_agent/__main__.py", line 123, in run_agent
    use_shell=use_shell)
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/contextlib.py", line 24, in __exit__
    self.gen.next()
  File "/usr/local/lib/python2.7/site-packages/trezor_agent/server.py", line 88, in serve
    server.shutdown(socket.SHUT_RD)
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/socket.py", line 228, in meth
    return getattr(self._sock,name)(*args)
socket.error: [Errno 57] Socket is not connected
^C^C^C^C^C^C^Z
[1]  + 5533 suspended  trezor-agent -vvvvv -c username@server.example.com
 ✘ ⚙ username@workstation > ~ > kill %1
[1]  + 5533 terminated  trezor-agent -vvvvv -c username@server.example.com
 ⚙ username@workstation > ~ >
```

**Comment by romanz** [comment]
Thanks for reporting this!
I will take a look.

**Comment by romanz** [comment]
@opie4624 BTW, does this bug happen all the time?
If so, could you please use https://asciinema.org/ to record your session when this issue takes place?

**Comment by opie4624** [comment]
@romanz Yeah, this happens every single time.
[![asciicast](https://asciinema.org/a/cmlh57afbvjmkxe8upir6utf3.png)](https://asciinema.org/a/cmlh57afbvjmkxe8upir6utf3)

**Comment by romanz** [comment]
Thanks for the detailed screencast!

Does the Trezor ask you to confirm the SSH login?
(it seems that you are using SSH session multiplexing: https://en.wikibooks.org/wiki/OpenSSH/Cookbook/Multiplexing)

**Comment by opie4624** [comment]
No, the Trezor never asks me to confirm the login. It only prompts me for the pin.
I am using multiplexing, but I checked that no multiplexing connections were open before I started the connections. I believe the multiplexing delay should be around 30s. But even so, the trezor-agent hangs around for much longer than that. (As an experiment, I left it in that state when I left work last night and it was still there this morning.)

**Comment by romanz** [comment]
This is interesting...

Can you try the following command:

    trezor-agent akraut@uhacc.club -vv -- ssh -o "ControlMaster=no" akraut@uhacc.club

**Comment by romanz** [comment]
Also, it seems that in OSX, calling `socket.shutdown()` on an accepting socket does not cause it to disconnect (it was the trick I have using in Linux to stop the "ssh-agent" serving thread after the SSH connection is closed).

**Comment by opie4624** [comment]
[![asciicast](https://asciinema.org/a/8w1q5xgcmix5bbefzmfz8nbxc.png)](https://asciinema.org/a/8w1q5xgcmix5bbefzmfz8nbxc)

**Comment by opie4624** [comment]
Also, just to be complete:
```akraut@uhacc:~$ ssh -V
OpenSSH_6.6.1p1 Ubuntu-2ubuntu2.3, OpenSSL 1.0.1f 6 Jan 2014```

**Comment by romanz** [comment]
Are you running `trezor-agent` on Ubuntu or Mac OS?

**Comment by opie4624** [comment]
My workstation is Mac OS:
```
akraut@akraut-m01 > ~ > ssh -V
OpenSSH_6.2p2, OSSLShim 0.9.8r 8 Dec 2011
```

**Comment by romanz** [comment]
It seems that the local SSH client fails to parse the ECDSA public keys returned from the Trezor.

**Comment by opie4624** [comment]
Ok, using a newer SSH on the mac allows it to connect. The failure to exit is still present:
[![asciicast](https://asciinema.org/a/4wvr9hj2honb7q96vrv2uf6m6.png)](https://asciinema.org/a/4wvr9hj2honb7q96vrv2uf6m6)

**Comment by romanz** [comment]
Can you please try the following:

  trezor-agent -vv foo_bar > /tmp/pub
  ssh-keygen -e -f /tmp/pub

This should print something like:

2016-01-06 22:19:00,747 DEBUG        connected to Trezor D996AB205905E96F77DE6663                                                         [client.py:23]
2016-01-06 22:19:00,747 DEBUG        label    : romanz                                                                                    [client.py:24]
2016-01-06 22:19:00,747 DEBUG        vendor   : bitcointrezor.com                                                                         [client.py:25]
2016-01-06 22:19:00,748 DEBUG        version  : 1.3.4                                                                                     [client.py:28]
2016-01-06 22:19:00,748 DEBUG        revision : 3c82ed984d9f1a0bfdce9f0977416b28313ce696                                                  [client.py:29]
2016-01-06 22:19:00,759 DEBUG        parsed identity: {'path': None, 'host': 'foo_bar', 'proto': None, 'port': None, 'user': None}        [client.py:101]
2016-01-06 22:19:00,759 INFO         getting "ssh://foo_bar" public key (nist256p1) from Trezor...                                        [client.py:54]
2016-01-06 22:19:00,759 DEBUG        address string: '\x00\x00\x00\x00ssh://foo_bar'                                                      [client.py:123]
2016-01-06 22:19:04,757 DEBUG        fingerprint: 4b:47:38:e7:72:84:be:71:d6:fe:66:6f:0d:69:81:8c                                         [formats.py:141]
2016-01-06 22:19:04,758 INFO         disconnected from Trezor                                                                             [client.py:41]
---- BEGIN SSH2 PUBLIC KEY ----
Comment: "256-bit ECDSA, converted by roman@lmde from OpenSSH"
AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBBqjTVh0RyaBkIOea4
Ndxh5Jo4OCzRcgWg4ux1/kU6exs/RXq6Vmwi2fuDdhi6gSFigpGUrQNZQrDDOTBKcDA7k=

---- END SSH2 PUBLIC KEY ----

**Comment by romanz** [comment]
Great!

**Comment by opie4624** [comment]
```
2016-01-06 12:20:28,846 DEBUG        connected to Trezor 358EB5A28279FB09FDA78CE9                                                         [client.py:23]
2016-01-06 12:20:28,846 DEBUG        label    : akraut                                                                                    [client.py:24]
2016-01-06 12:20:28,846 DEBUG        vendor   : bitcointrezor.com                                                                         [client.py:25]
2016-01-06 12:20:28,846 DEBUG        version  : 1.3.4                                                                                     [client.py:28]
2016-01-06 12:20:28,846 DEBUG        revision : db93a50f76204418a2cf7d2c7e0391f486729bf3                                                  [client.py:29]
2016-01-06 12:20:28,862 DEBUG        parsed identity: {'path': None, 'host': 'foo_bar', 'proto': None, 'port': None, 'user': None}        [client.py:101]
2016-01-06 12:20:28,862 INFO         getting "ssh://foo_bar" public key (nist256p1) from Trezor...                                        [client.py:54]
2016-01-06 12:20:28,862 DEBUG        address string: '\x00\x00\x00\x00ssh://foo_bar'                                                      [client.py:123]
Use the numeric keypad to describe number positions. The layout is:
    7 8 9
    4 5 6
    1 2 3
Please enter current PIN:

2016-01-06 12:20:54,230 DEBUG        fingerprint: f8:5c:08:67:2c:1d:63:15:b7:21:1e:b4:c9:65:60:3f                                         [formats.py:141]
2016-01-06 12:20:54,230 INFO         disconnected from Trezor                                                                             [client.py:41]
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@         WARNING: UNPROTECTED PRIVATE KEY FILE!          @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
Permissions 0644 for '/tmp/pub' are too open.
It is required that your private key files are NOT accessible by others.
This private key will be ignored.
bad permissions: ignore key: /tmp/pub
Enter passphrase:
```

**Comment by romanz** [comment]
Regarding the failure after the logout - it seems it happens due to a difference Mac OS X and Linux.

A bit of background about SSH client is needed: if it runs with SSH_AUTH_SOCK environment variable, it tries to connect to this UNIX socket, and asks the server (called "ssh-agent") behind the socket to perform the actual signing on its behalf.

In our case, I am using the same interface to forward the signing request to the Trezor device and replying with the correct response to the SSH client.
After the SSH client finishes, I have to stop also the "server" - and I did it by calling `socket.shutdown()` on its listening socket, which worked quite well on my Linux machine - but seems to fail on Mac OS X.

**Comment by romanz** [comment]
Can you please run the following Python code on your machine to reproduce the scenario of shutting down a listening UNIX socket server (before calling `accept()`, to make the code simpler):

```
import socket
s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
s.bind('/tmp/foobar.sock')
s.listen(1)
s.shutdown(socket.SHUT_RD)
```

**Comment by opie4624** [comment]
Looks like it needs an accept in order to properly shutdown? My Python Socket Foo is very weak at the moment.

https://asciinema.org/a/3lvdml0gwqzkzgnp2o9a48wht

**Comment by romanz** [comment]
Well, it seems that the socket system calls behave differently on Mac OS X and Linux...
https://asciinema.org/a/3j3pq1gn4m8mncm0hy67lpi64

**Comment by romanz** [comment]
I have a few ideas how to make it work, but the main problem is that I don't have any computer running Mac OS under my control... :(

**Comment by romanz** [comment]
I will mark this issue with "help wanted" label, for any one that have access to Mac OS and a Trezor wallet and can help us to fix this issue.

**Comment by romanz** [comment]
On the other hand, you can try the following patch (just an ugly hack to stop the hanging of `trezor-agent` after the SSH session is over):
```diff
diff --git a/trezor_agent/server.py b/trezor_agent/server.py
index be81d91..cb19b0a 100644
--- a/trezor_agent/server.py
+++ b/trezor_agent/server.py
@@ -66,6 +66,7 @@ def server_thread(server, handler):
 @contextlib.contextmanager
 def spawn(func, **kwargs):
     t = threading.Thread(target=func, kwargs=kwargs)
+    t.daemon = True
     t.start()
     yield
     t.join()
```

**Comment by romanz** [comment]
Please try the latest version of this library (from https://github.com/romanz/trezor-agent/commit/fb0d0a5f6138d3e55786ae8b32b6ea2d231527c9). 
I have replace the incompatible socket system call, by a more generic polling mechanism that should work on either platform (adding up to 100ms latency for closing the program).

**Comment by romanz** [comment]
@opie4624 Did you try the latest version on Mac OS?

**Comment by romanz** [comment]
You can also now set the timeout via a command-line argument:
https://github.com/romanz/trezor-agent/commit/b101281a5b53477d3553bd4b9a01cb38e513537e

**Comment by romanz** [comment]
Please take a look on the latest code - we have fixed another bug on Mac OS X (#9).

**Comment by romanz** [comment]
@opie4624 are there any news about the bug?
